File: sdks/java/src/sqlflow-core/src/main/java/de/bytefish/sqlflow/core/workers/GenericSqlFlowWorker.java
package de.bytefish.sqlflow.core.workers;

import de.bytefish.sqlflow.core.SqlFlow;
import de.bytefish.sqlflow.core.infrastructure.JobFactory;
import de.bytefish.sqlflow.core.infrastructure.SqlFlowRegistry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class GenericSqlFlowWorker implements AutoCloseable {
    private static final Logger logger = LoggerFactory.getLogger(GenericSqlFlowWorker.class);
    private final SqlFlow client;
    private final SqlFlowRegistry registry;
    private final JobFactory jobFactory;
    private final String queueName;
    private SqlFlowWorker innerWorker;

    public GenericSqlFlowWorker(SqlFlow client, SqlFlowRegistry registry, JobFactory jobFactory, String queueName) {
        this.client = client;
        this.registry = registry;
        this.jobFactory = jobFactory;
        this.queueName = queueName;
    }

    public void start() {
        logger.info("Initializing worker for queue: {}", queueName);
        client.createQueue(queueName);

        WorkerConfiguration config = registry.getWorkerConfigs().stream()
                .filter(c -> c.queueName().equals(queueName))
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("No config found for queue: " + queueName));

        // Register handlers for this queue
        var registrations = registry.getJobRegistrationsByQueue().get(queueName);
        if (registrations != null) {
            for (var reg : registrations) {
                reg.register(client, jobFactory);
            }
        }

        WorkerOptions options = WorkerOptions.builder()
                .workerId("worker-" + queueName + "-" + System.nanoTime())
                .queue(queueName)
                .claimTimeout(config.claimTimeoutInSeconds())
                .batchSize(config.batchSize() != null ? config.batchSize() : config.concurrency())
                .concurrency(config.concurrency())
                .pollInterval(config.pollIntervalInSeconds())
                .onError(config.onError() != null ? config.onError() : ex -> logger.error("Worker error", ex))
                .fatalOnLeaseTimeout(config.fatalOnLeaseTimeout())
                .build();

        SqlFlowWorker worker = new SqlFlowWorker(options, client);
        this.innerWorker = worker;
        
        Thread.ofVirtual().start(worker::run);
    }

    @Override
    public void close() throws Exception {
        if (this.innerWorker != null) {
            this.innerWorker.close();
        }
    }
}

File: sdks/dotnet/src/SqlFlowSdk/Workers/GenericSqlFlowWorker.cs
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using SqlFlowSdk.Core;

namespace SqlFlowSdk.Workers;

internal sealed class GenericSqlFlowWorker : BackgroundService
{
    private readonly ISqlFlow _client;
    private readonly ISqlFlowDispatcher _dispatcher;
    private readonly IServiceProvider _provider;
    private readonly SqlFlowRegistry _registry;
    private readonly ILogger<GenericSqlFlowWorker> _logger;
    private readonly string _queueName;

    public GenericSqlFlowWorker(
        ISqlFlow client,
        ISqlFlowDispatcher dispatcher,
        IServiceProvider provider,
        SqlFlowRegistry registry,
        ILogger<GenericSqlFlowWorker> logger,
        string queueName)
    {
        _client = client;
        _dispatcher = dispatcher;
        _provider = provider;
        _registry = registry;
        _logger = logger;
        _queueName = queueName;
    }

    protected override async Task ExecuteAsync(
        CancellationToken stoppingToken)
    {
        WorkerConfiguration config = _registry.WorkerConfigs.Single(x => x.QueueName == _queueName);

        _logger.LogInformation(
            "Creating queue {Queue} if it does not exist.",
            _queueName);

        await _client.CreateQueueAsync(
                _queueName,
                stoppingToken)
            .ConfigureAwait(false);

        await RegisterTasksAsync().ConfigureAwait(false);

        WorkerOptions options = new()
        {
            Queue = config.QueueName,
            WorkerId = CreateWorkerId(config.QueueName),
            Concurrency = config.Concurrency,
            BatchSize = config.BatchSize,
            ClaimTimeout = config.ClaimTimeoutInSeconds,
            FatalOnLeaseTimeout = config.FatalOnLeaseTimeout,
            OnError = config.OnError ??
                (exception => _logger.LogError(
                    exception,
                    "Worker error in queue {Queue}.",
                    config.QueueName))
        };

        _logger.LogInformation(
            "Starting worker for queue {Queue} with concurrency {Concurrency}.",
            options.Queue,
            options.Concurrency);

        await _dispatcher.RunWorkerAsync(
                options,
                stoppingToken)
            .ConfigureAwait(false);
    }

    private async Task RegisterTasksAsync()
    {
        if (!_registry.JobRegistrationsByQueue.TryGetValue(
                _queueName,
                out var registrations))
        {
            return;
        }

        foreach (var registration in registrations)
        {
            await registration(_client, _provider)
                .ConfigureAwait(false);
        }
    }

    private static string CreateWorkerId(
        string queueName)
    {
        string suffix = Guid.NewGuid()
            .ToString("N")[..6];

        return $"worker-{queueName}-{suffix}";
    }
}